text/0000-generic-dependent-consts.md

```diff
@@ -0,0 +1,291 @@
+- Feature Name: generic_dependent_consts
+- Start Date: 2015-03-28
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Allow consts declared in functions to depend on function type parameters. Since
+this raises some issues in match and type checks, use an conservative approach
+that allows only minimal use of such consts when a constant expression is
+required.
+
+# Motivation
+
+Consider the following line of code in a non-generic context:
+
+```rust
+let a: [u8; T::N + T::N] = [0u8; 2*T::N];
+```
+
+In the current associated consts implementation (modulo some temporary wrinkles
+involving constants associated to traits), this will readily type-check. Since
+the type `T` must be known in non-generic code, `T::N` will evaluate to some
+value, say `16`, in which case it is clear that both sides of the assignment
+involve arrays of size `32`.
+
+However, in generic code, if `T` is a type parameter, we are asking the type
```

Comment:
@tommit:

Don't worry about it. I know that it these discussions cause some clutter, and I hope that that doesn't discourage people from reading, but at the same time I had almost exactly the same reaction as you only a few months ago, when I was still thinking of Rust generics as being like C++ templates (even though I thought I had absorbed enough from Haskell, maybe ML, to know better). So my hope is that people reading are either interested enough to get through it, ADHD enough to skip to the next part, or green enough that it's educational for them.

@freebroccolo:

Thanks for your explanation (as you may have guessed, I mentioned you in the hope of getting you involved here!). I think that in a very broad sense, I had the gist of what you intended, but I'll have to consider the implementation details you mentioned, which I either didn't see before or simply forgot.

I'll have to consider this when I have the time to do it justice. In the meantime, I'd appreciate your thoughts on this RFC in general, and particularly if it's compatible with the route you'd take, whether or not it lines up conceptually. (Preferably on the main thread rather than this line note, which is getting a bit excessive; admittedly that's my fault.)